### PR TITLE
PDF/A 3b

### PR DESCRIPTION
**Pull request by ilaurillard**
Would you please have a look at the catalog.dart file - there is a problem with the "/Names" part.
You have an indirect/ref field and i have a "PdfDict" - these have to be merged somehow ..

greetings

### CODE REVIEW

File: pdf/lib/src/pdf/obj/pdf_a_3b.dart
Comment (DavBfr) [inline pdf/lib/src/pdf/obj/pdf_a_3b.dart:1]: Perhaps this file could be divided into multiple class files.

File: pdf/lib/src/pdf/obj/catalog.dart [hunk 1]
```diff
@@ -94,6 +101,12 @@ class PdfCatalog extends PdfObject<PdfDict> {
       params['/Names'] = names!.ref();
     }
 
+    // ??? what to do, if /Names is already occupied?
+    if (attached != null && attached!.isNotEmpty) {
+      params['/Names'] = attached!.catalogNames();
```
Comment (DavBfr) [inline pdf/lib/src/pdf/obj/catalog.dart:106]: If names already exists, just merge the dict here.

```dart
if (params.containsKey('/Names')) {
  params['/Names']!.merge(attached!.catalogNames());
} else {
  params['/Names'] = attached!.catalogNames();
}
```

**Review by DavBfr — CHANGES_REQUESTED** [review]

File: pdf/lib/src/pdf/obj/annotation.dart [hunk 2]
```diff
@@ -299,6 +299,12 @@ abstract class PdfAnnotBase {
     if (flags != null && flags!.isNotEmpty) {
       params['/F'] = PdfNum(flagValue);
     }
+    else {
```
[the remark this answers is not in the document]
Comment (ilaurillard) [reply pdf/lib/src/pdf/obj/annotation.dart]: ok, made attribute in base class non nullable - without changing derived classes - we need to make this late final, or?

File: pdf/lib/src/pdf/obj/pdf_a_3b.dart
Comment (ilaurillard) [reply pdf/lib/src/pdf/obj/pdf_a_3b.dart:1]: ok, done

File: pdf/lib/src/pdf/obj/catalog.dart (hunk 1 above)
Comment (ilaurillard) [reply pdf/lib/src/pdf/obj/catalog.dart]: since names is a ref, i know merge the pdfa-names into the pdfNames .. ok?

**Comment by grundid** [comment]
@ilaurillard Thank you for your pdf/a implementation. I see you have added some code related to XRechnung/Factur-X. I'm also working on an implementation for e-invoices.
Could you please make ```PdfaFacturxRdf``` more customizable? 
It should be possible to set the ```ConformanceLevel``` and not only fix it to "BASIC".

At the same it might be better to put this code in a separated library as it is more related to e-invoices than generic PDF generation. I could provide my schema dart code for e-invoices.

**Comment by ilaurillard** [comment]
> @ilaurillard Thank you for your pdf/a implementation. I see you have added some code related to XRechnung/Factur-X. I'm also working on an implementation for e-invoices. Could you please make `PdfaFacturxRdf` more customizable? It should be possible to set the `ConformanceLevel` and not only fix it to "BASIC".
> 

done

> At the same it might be better to put this code in a separated library as it is more related to e-invoices than generic PDF generation. I could provide my schema dart code for e-invoices.

yes, you are right..  pdf/a and e-invoices are separate things ... 
this should be discussed .. @DavBfr?

@grundid  for UBL invoices i wrote a class generator, that crawls the official xsd schemas an makes dart classes of the definitions, if you are interested ..

**Comment by grundid** [comment]
> @grundid for UBL invoices i wrote a class generator, that crawls the official xsd schemas an makes dart classes of the definitions, if you are interested ..

Well, yes! Right now I've done the same thing for the ZUGFeRD specification in Germany and now I'm validating the output xml agains the schematron rules. 
Shall we set up a project and discuss collaboration?

**Comment by DavBfr** [comment]
Yes, all schema generation could be in another package.

**Comment by grundid** [comment]
I've created a dart package dedicated to the creation of e-invoices based on the ZUGFeRD/Factur-X specification: https://github.com/Blaulichtplaner/e_invoice_tools

@ilaurillard you can find an example with all the schema classes here:
https://github.com/Blaulichtplaner/e_invoice_tools/blob/main/packages/e_invoice_data_classes/example/e_invoice_data_classes_example.dart

File: pdf/lib/src/pdf/obj/pdfa/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+Here are some classes to help you creating PDF/A compliant PDFs
+plus embedding Facturx invoices.
+
+### Rules
+
+1. Your PDF must only use embedded Fonts,
+2. For now you cannot use any Annotations in your PDF
+3. You must include a special Meta-XML, use below "PdfaRdf" and put the reuslting XML document into your documents metadata
+4. You muss include a Colorprofile, use the below "PdfaColorProfile" and embed the contents of "sRGB2014.icc"
```
Comment (CLNMR) [inline pdf/lib/src/pdf/obj/pdfa/README.md:9]:
```suggestion
4. You must include a Colorprofile, use the below "PdfaColorProfile" and embed the contents of "sRGB2014.icc"
```

**Comment by CLNMR** [comment]
Hi, is there any timeline when this could be merged? I need the capability to attach files (xml-documents) to my pdf to create a valid ZUGFeRD document :)

**Review by DavBfr — APPROVED** [review]

**Pull request merged**